sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -4600,3 +4600,51 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
   override protected def withNewChildrenInternal(
     newLeft: Expression, newRight: Expression): ArrayExcept = copy(left = newLeft, right = newRight)
 }
+
+@ExpressionDescription(
+  usage = "_FUNC_(array) - Removes null values from the array.",
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3, null));
+       [1,2,3]
+  """,
+  group = "array_funcs",
+  since = "3.4.0")
+case class ArrayCompact(child: Expression)
```

Comment:
if the input is `Array(Array(1,null,3), null, Array(null, 2, 3))`, what is the result?

Reply:
```
scala>     val df_test = Seq(Seq(Array[Integer](1, null, 3), null, Array[Integer](null, 2, 3))).toDF("a")
df_test: org.apache.spark.sql.DataFrame = [a: array<array<int>>]

scala>     df_test.select(array_compact($"a")).show(false)
+----------------------------+
|array_compact(a)            |
+----------------------------+
|[[1, null, 3], [null, 2, 3]]|
+----------------------------+

```
I guess it is wrong, I believe I also need to traverse the element if it is of type Array. Great find!!!

Reply:
I personally think the output `[[1, null, 3], [null, 2, 3]]` is expected, let me confirm it.

Reply:
@zhengruifeng I think it is better to give some typical cases in jira. Personally, the document description of snowflake is not very clear :(

Reply:
I have updated the Unit Tests to reflect this behaviour. If we required to change then I will update the code and UT

Reply:
Got it, thanks

Reply:
> I personally think the output `[[1, null, 3], [null, 2, 3]]` is expected, let me confirm it.

Is there a clear conclusion?

Reply:
I think we can leave the current behaviour as we have `transform` API which can be used in such cases 
for e.g
```
scala> val df_test = Seq(Seq(Array[Integer](1, null, 3), null, Array[Integer](null, 2, 3))).toDF("a")
df_test: org.apache.spark.sql.DataFrame = [a: array<array<int>>]

scala> df_test.select(array_compact(transform($"a", x => array_compact(x))).as("t")).show(false)
+----------------+
|t               |
+----------------+
|[[1, 3], [2, 3]]|
+----------------+
```

Reply:
In this case, I think we do not need to exactly match Snowflake.

I think we should refer to `ARRAY_DISTINCT` which will not deduplicate its elements when its elements are also arrays:

```
scala> spark.sql(""" SELECT array_distinct(array(a,b,c)) FROM VALUES (ARRAY(1,1,2), ARRAY(8,8), ARRAY(1,1,2)) AS tab(a,b,c) """).head
res19: org.apache.spark.sql.Row = [WrappedArray(WrappedArray(1, 1, 2), WrappedArray(8, 8))]
```

also cc @HyukjinKwon @cloud-fan

Reply:
+1 for @zhengruifeng for now

Reply:
```
select array_compact(array_construct(array_construct(1, null), null));
+----------------------------------------------------------------+
| ARRAY_COMPACT(ARRAY_CONSTRUCT(ARRAY_CONSTRUCT(1, NULL), NULL)) |
|----------------------------------------------------------------|
| [                                                              |
|   [                                                            |
|     1,                                                         |
|     undefined                                                  |
|   ]                                                            |
| ]                                                              |
+----------------------------------------------------------------+
1 Row(s) produced. Time Elapsed: 0.112s
```

FYI, we confirmed that snowflake's `array_compact` will not remove its element arrays' NULL values.